File: apps/api/app/schemas/role.py
"""Role schemas."""

from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import Field

from app.schemas.base import BaseSchema
from app.schemas.permission import PermissionRead


class RoleBase(BaseSchema):
    name: str = Field(min_length=1, max_length=100)
    description: str | None = None


class RoleCreate(RoleBase):
    permission_ids: list[uuid.UUID] = Field(default_factory=list)


class RoleUpdate(BaseSchema):
    name: str | None = Field(default=None, min_length=1, max_length=100)
    description: str | None = None


class RoleRead(BaseSchema):
    id: uuid.UUID
    name: str
    slug: str
    description: str | None
    is_system: bool
    permissions: list[PermissionRead]
    created_at: datetime
    updated_at: datetime


class RoleReadSlim(BaseSchema):
    id: uuid.UUID
    name: str
    slug: str
    description: str | None
    is_system: bool


class AssignPermissionsRequest(BaseSchema):
    permission_ids: list[uuid.UUID]


class AssignRolesRequest(BaseSchema):
    role_ids: list[uuid.UUID]

File: apps/web/src/types/role.ts
import type { PermissionRead } from './permission';

export interface RoleRead {
  id: string;
  name: string;
  slug: string;
  description: string | null;
  isSystem: boolean;
  permissions: PermissionRead[];
  createdAt: string;
  updatedAt: string;
}

export interface RoleReadSlim {
  id: string;
  name: string;
  slug: string;
  description: string | null;
  isSystem: boolean;
}

export interface RoleCreate {
  name: string;
  description?: string;
  permissionIds?: string[];
}

export interface RoleUpdate {
  name?: string;
  description?: string;
}

export interface AssignPermissionsRequest {
  permissionIds: string[];
}

export interface AssignRolesRequest {
  roleIds: string[];
}
